sips/verifiable_consensus_data.md

```diff
@@ -0,0 +1,235 @@
+                                                     sourceProof phase0.SignedBeaconBlockHeader, 
+                                                     error) {
+    if err := att.UnmarshalSSZ(cd.DataSSZ); err != nil {
+        return nil, nil, nil, nil, errors.Wrap(err, "could not unmarshal attestation ssz")
+    }
+    // Head vote proof
+    if err := headProof.UnmarshalSSZ(cd.Proof[0]); err != nil {
+        return nil, nil, errors.Wrap(err, "could not unmarshal head vote proof ssz")
+    }
+    // Target proof
+    if err := targetProof.UnmarshalSSZ(cd.Proof[1]); err != nil {
+        return nil, nil, errors.Wrap(err, "could not unmarshal target proof ssz")
+    }
+    // Target proof
+    if err := sourceProof.UnmarshalSSZ(cd.Proof[1]); err != nil {
+        return nil, nil, errors.Wrap(err, "could not unmarshal source proof ssz")
+    }
+    return att, headProof, targetProof, sourceProof, nil
+}
+
+func (cd VerifiableConsensusData) GetSyncCommitteeBlockRoot (phase0.Root, phase0.SignedBeaconBlockHeader, error) {
+	ret := SSZ32Bytes{}
+	if err := ret.UnmarshalSSZ(cd.DataSSZ); err != nil {
+		return nil, errors.Wrap(err, "could not unmarshal ssz")
+	}
+    proof := phase0.SignedBeaconBlockHeader
+    if err := proof.UnmarshalSSZ(cd.Proof); err != nil {
+        return nil, nil, errors.Wrap(err, "could not unmarshal proof ssz")
+    }
+	return phase0.Root(ret), proof, nil
+}
+
+func (cd VerifiableConsensusData) Validate error {
+    switch cid.Duty.Type {
+	case BNRoleAttester:
+            if _, _, _, _, err := cid.GetAttestationData(); err != nil {
+                return err
+            }
+            return nil
+	case BNRoleSyncCommittee:
+            if _, _, err := cid.GetSyncCommitteeBlockRoot(); err != nil {
+                return err
+            }
+            return nil
+        default:
+		return errors.New("unknown duty role")
+	}
+}
+```
+
+In addition we do the following changes to value checks mechanisms:
+```go 
+func verifyData(expectedRoot phase0.Root, proof phase0.SignedBeaconBlockHeader) error {
+    beaconRoot := proof.Message.HashRoot()
+    if !bytes.Equal(expectedRoot, beaconRoot) {
+        return errors.New("unexpected header root")
+    }
+    blockEpoch := getBeaconNode().getEpochBySlot(proof.Message.slot)
+    domain := GetBeaconNode().DomainData(blockEpoch, spec.DomainProposer)
+    root := spec.ComputeEthSigningRoot(beaconRoot, domain)
+    validator := GetBeaconNode.GetValidator("HEAD", proof.Message.ProposerIndex)
+    return bls.Verify(sig, validator.PubKey, root)
+}
+
+func AttesterValueCheckF(
+	signer types.BeaconSigner,
+	network types.BeaconNetwork,
+	validatorPK types.ValidatorPK,
+	validatorIndex phase0.ValidatorIndex,
+	sharePublicKey []byte,
+    // obtained from the operator's beacon node
+    attestationDataByts []byte,
+) qbft.ProposedValueCheckF {
+	return func(data []byte) error {
+        // addition
+        if bytes.Equal(attestationDataByts, data) {
+            return nil
+        }
```

Comment:
This is a `nil` or empty check right?

Reply:
if the attestation data we got from our own node is the same as the leader's (it should happen if the nodes are synced to the same head) then no need to perform any more validation and we stop

This assumes the operator trusts his beacon node (most of the time they run their own I think).

If the `data` is `nil` it should fail I think when we do `deocode` or `validate`